File: webui/src/types/workflow.ts
export interface Workflow {
  id: string;
  name: string;
  description: string;
  workflow_type: string;
  activity: ActivityType;
  params: Record<string, unknown>;
  target_filter: string;
  trigger: TriggerSpec;
  success_threshold: number;
  loop_threshold_ms: number;
  timeout_ms: number;
  active: boolean;
  created_at: string;
  updated_at: string;
  deactivated_reason?: string;
}

export type ActivityType =
  | 'reboot'
  | 'install_package'
  | 'upgrade_package'
  | 'remove_package'
  | 'apply_config'
  | 'validate_config'
  | 'run_script';

export type TriggerKind = 'scheduled' | 'event' | 'state_change' | 'manual';

export interface TriggerSpec {
  kind: TriggerKind;
  cron?: string;
  on_complete?: string;
  on_state_change?: string;
}

export interface CreateWorkflowRequest {
  name: string;
  description: string;
  workflow_type: string;
  activity: ActivityType;
  params: Record<string, unknown>;
  target_filter: string;
  trigger: TriggerSpec;
  success_threshold: number;
  loop_threshold_ms: number;
  timeout_ms: number;
  active: boolean;
}

export interface EditWorkflowRequest extends CreateWorkflowRequest {}

export interface TriggerWorkflowRequest {
  reason?: string;
}

File: internal/core/domain/workflow.go
package domain

import (
	"time"
)

type TriggerKind string

const (
	TriggerScheduled   TriggerKind = "scheduled"
	TriggerEvent       TriggerKind = "event"
	TriggerStateChange TriggerKind = "state_change"
	TriggerManual      TriggerKind = "manual"
)

type TriggerSpec struct {
	Kind       TriggerKind            `json:"kind"`
	Cron       string                 `json:"cron,omitempty"`
	OnComplete string                 `json:"on_complete,omitempty"`
	Condition  string                 `json:"condition,omitempty"`
	Params     map[string]interface{} `json:"params,omitempty"`
}

type Workflow struct {
	ID                string                 `json:"id"`
	Name              string                 `json:"name"`
	Description       string                 `json:"description"`
	WorkflowType      string                 `json:"workflow_type"`
	Activity          ActivityType           `json:"activity"`
	Params            map[string]interface{} `json:"params"`
	TargetFilter      string                 `json:"target_filter"`
	Trigger           TriggerSpec            `json:"trigger"`
	SuccessThreshold  float64                `json:"success_threshold"`
	LoopThresholdMS   int64                  `json:"loop_threshold_ms"`
	TimeoutMS         int64                  `json:"timeout_ms"`
	Active            bool                   `json:"active"`
	CreatedAt         time.Time              `json:"created_at"`
	UpdatedAt         time.Time              `json:"updated_at"`
	DeactivatedReason string                 `json:"deactivated_reason,omitempty"`
}

func (w *Workflow) IsActive() bool { return w.Active }

func (w *Workflow) ValidateDefinition() error {
	if w.Name == "" {
		return ErrMissingRequiredField
	}
	if w.WorkflowType == "" {
		return ErrMissingRequiredField
	}
	if !w.Activity.IsValid() {
		return ErrInvalidWorkflow
	}
	if w.SuccessThreshold < 0 || w.SuccessThreshold > 100 {
		return ErrInvalidWorkflow
	}
	if w.LoopThresholdMS < 0 {
		return ErrInvalidWorkflow
	}
	if w.TimeoutMS < 0 {
		return ErrInvalidWorkflow
	}
	return nil
}

func (w *Workflow) GetActivityTimeout() time.Duration {
	if w.TimeoutMS <= 0 {
		return 30 * time.Second
	}
	return time.Duration(w.TimeoutMS) * time.Millisecond
}

func (w *Workflow) LoopThreshold() time.Duration {
	if w.LoopThresholdMS <= 0 {
		return 5 * time.Second
	}
	return time.Duration(w.LoopThresholdMS) * time.Millisecond
}
